File: tui/agent_runtime.py
from __future__ import annotations

import os
import re
import shutil
import subprocess
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Protocol


FLAG_PATTERN = re.compile(r"FLAG\{[A-Za-z0-9_/-]+\}", re.IGNORECASE)
LOCAL_HOSTS = {"127.0.0.1", "localhost", "::1"}


class FlagSubmitter(Protocol):
    def __call__(self, flag: str, target_url: str) -> dict[str, Any]: ...


@dataclass(slots=True)
class AgentContext:
    match_id: str
    room_id: str
    team_id: str
    phase: str
    targets: list[dict[str, Any]]
    local_target: dict[str, Any]
    allowed_targets: list[str]
    target_template_id: str
    timeout_sec: int = 300

    def env(self) -> dict[str, str]:
        return {
            "AIAWD_MATCH_ID": self.match_id,
            "AIAWD_ROOM_ID": self.room_id,
            "AIAWD_TEAM_ID": self.team_id,
            "AIAWD_PHASE": self.phase,
            "AIAWD_TARGETS": json_dumps([t["base_url"] for t in self.targets]),
            "AIAWD_LOCAL_TARGET": self.local_target.get("base_url", ""),
            "AIAWD_ALLOWED_TARGETS": json_dumps(self.allowed_targets),
            "AIAWD_TARGET_TEMPLATE": self.target_template_id,
        }


@dataclass(slots=True)
class AgentAction:
    timestamp: float
    action: str
    target_url: str | None
    flag: str | None
    output: str
    ok: bool


@dataclass(slots=True)
class AgentResult:
    ok: bool
    actions: list[AgentAction] = field(default_factory=list)
    flags_captured: list[str] = field(default_factory=list)
    error: str | None = None
    elapsed_ms: int = 0


class AgentAdapter(ABC):
    @abstractmethod
    def configure(self, ctx: AgentContext) -> None: ...

    @abstractmethod
    def run(self, *, submit: FlagSubmitter | None = None) -> AgentResult: ...

    @abstractmethod
    def stop(self) -> None: ...


class CustomCommandAdapter(AgentAdapter):
    def __init__(
        self,
        command: list[str],
        *,
        cwd: Path | None = None,
        env: dict[str, str] | None = None,
        timeout_sec: int = 300,
        extract_flags: Callable[[str], list[str]] | None = None,
    ) -> None:
        self._command_template = command
        self._cwd = cwd or Path.cwd()
        self._extra_env = env or {}
        self._timeout = timeout_sec
        self._extract_flags = extract_flags or _default_extract_flags
        self._ctx: AgentContext | None = None
        self._process: subprocess.Popen[bytes] | None = None

    def configure(self, ctx: AgentContext) -> None:
        self._ctx = ctx
        self._timeout = ctx.timeout_sec

    def run(self, *, submit: FlagSubmitter | None = None) -> AgentResult:
        if self._ctx is None:
            return AgentResult(ok=False, error="Agent 未配置比赛上下文")
        started = time.time()
        actions: list[AgentAction] = []
        captured: list[str] = []
        targets = self._ctx.targets
        for target in targets:
            if not target:
                continue
            url = target.get("base_url", "")
            if not url:
                continue
            action = self._run_against(url, submit=submit)
            actions.append(action)
            if action.flag:
                captured.append(action.flag)
            if not action.ok:
                break
        elapsed = int((time.time() - started) * 1000)
        return AgentResult(
            ok=all(a.ok for a in actions),
            actions=actions,
            flags_captured=captured,
            elapsed_ms=elapsed,
        )

    def _run_against(self, target_url: str, *, submit: FlagSubmitter | None) -> AgentAction:
        argv = _expand_template(self._command_template, target_url=target_url, ctx=self._ctx)
        env = {**os.environ, **(self._ctx.env() if self._ctx else {}), **self._extra_env}
        started = time.time()
        try:
            proc = subprocess.run(
                argv,
                cwd=self._cwd,
                env=env,
                capture_output=True,
                text=True,
                timeout=self._timeout,
                check=False,
                shell=False,
            )
        except subprocess.TimeoutExpired:
            return AgentAction(
                timestamp=started,
                action="attack",
                target_url=target_url,
                flag=None,
                output="",
                ok=False,
            )
        output = (proc.stdout or "") + (proc.stderr or "")
        flags = self._extract_flags(output)
        for flag in flags:
            if submit:
                submit(flag, target_url)
        return AgentAction(
            timestamp=started,
            action="attack",
            target_url=target_url,
            flag=flags[0] if flags else None,
            output=output,
            ok=True,
        )

    def stop(self) -> None:
        if self._process and self._process.poll() is None:
            self._process.terminate()
            try:
                self._process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._process.kill()
                self._process.wait()


class AgentManager:
    def __init__(self, adapter: AgentAdapter) -> None:
        self.adapter = adapter
        self._results: list[AgentResult] = []
        self._running = False

    def configure(self, match_config: dict[str, Any], room_status: str) -> None:
        targets = match_config.get("opponents") or []
        ctx = AgentContext(
            match_id=match_config.get("match_id", ""),
            room_id=match_config.get("room_id", ""),
            team_id=match_config.get("team_id", ""),
            phase=room_status,
            targets=targets,
            local_target=match_config.get("local_target") or {},
            allowed_targets=match_config.get("allowed_targets") or [],
            target_template_id=match_config.get("target_template_id", ""),
        )
        self.adapter.configure(ctx)

    def run_attack(self, *, submit: FlagSubmitter | None = None) -> AgentResult:
        self._running = True
        result = self.adapter.run(submit=submit)
        self._results.append(result)
        self._running = False
        return result

    def stop(self) -> None:
        self.adapter.stop()
        self._running = False

    @property
    def running(self) -> bool:
        return self._running

    @property
    def last_result(self) -> AgentResult | None:
        return self._results[-1] if self._results else None


def _default_extract_flags(text: str) -> list[str]:
    return FLAG_PATTERN.findall(text)


def _expand_template(template: list[str], *, target_url: str, ctx: AgentContext | None) -> list[str]:
    replacements = {
        "{target_url}": target_url,
        "{local_target}": ctx.local_target.get("base_url", "") if ctx else "",
        "{match_id}": ctx.match_id if ctx else "",
        "{room_id}": ctx.room_id if ctx else "",
        "{team_id}": ctx.team_id if ctx else "",
    }
    result: list[str] = []
    for token in template:
        for placeholder, value in replacements.items():
            token = token.replace(placeholder, value)
        result.append(token)
    return result


def json_dumps(obj: Any) -> str:
    import json as _json

    return _json.dumps(obj, ensure_ascii=False, separators=(",", ":"))


def check_docker_available() -> bool:
    return shutil.which("docker") is not None


_SHELL_CONTROL = {";", "&", "&&", "||", "|"}
_SHELL_EXPANSION = ["$(", "${", "`"]
_SHELL_DANGEROUS = {";", "\n", "\r"}


def sanitize_command(command: list[str]) -> bool:
    for token in command:
        if token in _SHELL_CONTROL:
            return False
        for expansion in _SHELL_EXPANSION:
            if expansion in token:
                return False
        for char in token:
            if char in _SHELL_DANGEROUS:
                return False
    return True

File: client/agentRuntime.js
"use strict";

const { spawn, spawnSync } = require("child_process");
const path = require("path");

const FLAG_PATTERN = /FLAG\{[A-Za-z0-9_/-]+\}/gi;
const SHELL_CONTROL = new Set([";", "&", "&&", "||", "|"]);
const SHELL_EXPANSION = ["$(", "${", "`"];
const SHELL_DANGEROUS = new Set([";", "\n", "\r", "\t", "#", "~"]);

/**
 * @typedef {Object} AgentContext
 * @property {string} matchId
 * @property {string} roomId
 * @property {string} teamId
 * @property {string} phase
 * @property {Array<{teamId: string, baseUrl: string}>} targets
 * @property {{host: string, port: number, baseUrl: string}} localTarget
 * @property {string[]} allowedTargets
 * @property {string} targetTemplateId
 * @property {number} timeoutSec
 */

/**
 * @typedef {Object} AgentAction
 * @property {number} timestamp
 * @property {string} action
 * @property {string|null} targetUrl
 * @property {string|null} flag
 * @property {string} output
 * @property {boolean} ok
 */

/**
 * @typedef {Object} AgentResult
 * @property {boolean} ok
 * @property {AgentAction[]} actions
 * @property {string[]} flagsCaptured
 * @property {string|null} error
 * @property {number} elapsedMs
 */

/**
 * @callback FlagSubmitter
 * @param {string} flag
 * @param {string} targetUrl
 * @returns {{ok: boolean, code?: string}}
 */

function makeContext(matchConfig, roomStatus) {
  const opponents = matchConfig.opponents || [];
  const localTarget = matchConfig.local_target || {};
  const allowedTargets = matchConfig.allowed_targets || [];
  return {
    matchId: matchConfig.match_id || "",
    roomId: matchConfig.room_id || "",
    teamId: matchConfig.team_id || "",
    phase: roomStatus || matchConfig.phase || "",
    targets: opponents.map((o) => ({ teamId: o.team_id, baseUrl: o.base_url })),
    localTarget: {
      host: localTarget.host || "127.0.0.1",
      port: localTarget.port || 0,
      baseUrl: localTarget.base_url || "",
    },
    allowedTargets,
    targetTemplateId: matchConfig.target_template_id || "",
    timeoutSec: 300,
  };
}

function contextEnv(ctx) {
  return {
    AIAWD_MATCH_ID: ctx.matchId,
    AIAWD_ROOM_ID: ctx.roomId,
    AIAWD_TEAM_ID: ctx.teamId,
    AIAWD_PHASE: ctx.phase,
    AIAWD_TARGETS: JSON.stringify(ctx.targets.map((t) => t.baseUrl)),
    AIAWD_LOCAL_TARGET: ctx.localTarget.baseUrl,
    AIAWD_ALLOWED_TARGETS: JSON.stringify(ctx.allowedTargets),
    AIAWD_TARGET_TEMPLATE: ctx.targetTemplateId,
  };
}

function extractFlags(text) {
  const matches = text.match(FLAG_PATTERN);
  return matches || [];
}

function expandTemplate(template, targetUrl, ctx) {
  const replacements = {
    "{target_url}": targetUrl,
    "{local_target}": ctx ? ctx.localTarget.baseUrl : "",
    "{match_id}": ctx ? ctx.matchId : "",
    "{room_id}": ctx ? ctx.roomId : "",
    "{team_id}": ctx ? ctx.teamId : "",
  };
  return template.map((token) => {
    let result = token;
    for (const [placeholder, value] of Object.entries(replacements)) {
      result = result.replaceAll(placeholder, value);
    }
    return result;
  });
}

function sanitizeCommand(command) {
  for (const token of command) {
    if (SHELL_CONTROL.has(token)) return false;
    for (const expansion of SHELL_EXPANSION) {
      if (token.includes(expansion)) return false;
    }
    for (const char of token) {
      if (SHELL_DANGEROUS.has(char)) return false;
    }
  }
  return true;
}

function commandNeedsShell(command, platform = process.platform) {
  return platform === "win32" && /\.(cmd|bat)$/i.test(String(command || ""));
}

function quoteShellArg(arg) {
  const text = String(arg);
  if (!/[ \t"]/.test(text)) return text;
  return `"${text.replace(/"/g, '\\"')}"`;
}

function prepareSpawnArgv(argv, platform = process.platform) {
  if (!commandNeedsShell(argv[0], platform)) return argv;
  return argv.map(quoteShellArg);
}

class CustomCommandAdapter {
  /**
   * @param {string[]} command
   * @param {Object} [options]
   * @param {string} [options.cwd]
   * @param {Object<string,string>} [options.env]
   * @param {number} [options.timeoutSec]
   * @param {(text: string) => string[]} [options.extractFlagsFn]
   */
  constructor(command, options = {}) {
    this._commandTemplate = command;
    this._cwd = options.cwd || process.cwd();
    this._extraEnv = options.env || {};
    this._timeout = options.timeoutSec || 300;
    this._extractFlagsFn = options.extractFlagsFn || extractFlags;
    /** @type {AgentContext|null} */
    this._ctx = null;
    /** @type {ChildProcess|null} */
    this._process = null;
  }

  configure(ctx) {
    this._ctx = ctx;
    this._timeout = ctx.timeoutSec || 300;
  }

  /**
   * @param {FlagSubmitter} [submit]
   * @returns {AgentResult}
   */
  run(submit) {
    if (!this._ctx) {
      return { ok: false, actions: [], flagsCaptured: [], error: "Agent 未配置比赛上下文", elapsedMs: 0 };
    }
    const started = Date.now();
    /** @type {AgentAction[]} */
    const actions = [];
    /** @type {string[]} */
    const captured = [];
    const targets = this._ctx.targets;
    for (const target of targets) {
      if (!target || !target.baseUrl) continue;
      const action = this._runAgainst(target.baseUrl, submit);
      actions.push(action);
      if (action.flag) captured.push(action.flag);
      if (!action.ok) break;
    }
    return {
      ok: actions.every((a) => a.ok),
      actions,
      flagsCaptured: captured,
      elapsedMs: Date.now() - started,
      error: null,
    };
  }

  _runAgainst(targetUrl, submit) {
    const argv = expandTemplate(this._commandTemplate, targetUrl, this._ctx);
    const needsShell = commandNeedsShell(argv[0]);
    const shellArgv = prepareSpawnArgv(argv);
    const env = { ...process.env, ...contextEnv(this._ctx), ...this._extraEnv };
    const started = Date.now();
    try {
      const proc = spawnSync(shellArgv[0], shellArgv.slice(1), {
        cwd: this._cwd,
        env,
        timeout: this._timeout * 1000,
        encoding: "utf-8",
        stdio: "pipe",
        shell: needsShell,
      });
      if (proc.error) {
        return {
          timestamp: started, action: "attack", targetUrl,
          flag: null, output: proc.error.message || "", ok: false,
        };
      }
      const flags = this._extractFlagsFn(proc.stdout || "");
      if (submit) {
        for (const flag of flags) submit(flag, targetUrl);
      }
      return {
        timestamp: started, action: "attack", targetUrl,
        flag: flags[0] || null, output: proc.stdout || "", ok: true,
      };
    } catch (err) {
      return {
        timestamp: started, action: "attack", targetUrl,
        flag: null, output: err.message || "", ok: false,
      };
    }
  }

  _runAgainstAsync(targetUrl, submit) {
    const argv = expandTemplate(this._commandTemplate, targetUrl, this._ctx);
    const needsShell = commandNeedsShell(argv[0]);
    const shellArgv = prepareSpawnArgv(argv);
    const env = { ...process.env, ...contextEnv(this._ctx), ...this._extraEnv };
    const started = Date.now();
    return new Promise((resolve) => {
      const child = spawn(shellArgv[0], shellArgv.slice(1), {
        cwd: this._cwd,
        env,
        timeout: this._timeout * 1000,
        stdio: "pipe",
        shell: needsShell,
      });
      let stdout = "";
      let stderr = "";
      child.stdout.on("data", (d) => { stdout += d.toString(); });
      child.stderr.on("data", (d) => { stderr += d.toString(); });
      child.on("error", (err) => {
        resolve({
          timestamp: started, action: "attack", targetUrl,
          flag: null, output: err.message, ok: false,
        });
      });
      child.on("close", (code) => {
        const output = stdout + stderr;
        if (code !== 0) {
          resolve({
            timestamp: started, action: "attack", targetUrl,
            flag: null, output, ok: false,
          });
          return;
        }
        const flags = this._extractFlagsFn(output);
        if (submit) {
          for (const flag of flags) submit(flag, targetUrl);
        }
        resolve({
          timestamp: started, action: "attack", targetUrl,
          flag: flags[0] || null, output, ok: true,
        });
      });
    });
  }

  /**
   * Async version of run() — does NOT block the event loop.
   * Use this from Electron main process or any UI thread.
   * @param {FlagSubmitter} [submit]
   * @returns {Promise<AgentResult>}
   */
  async runAsync(submit) {
    if (!this._ctx) {
      return { ok: false, actions: [], flagsCaptured: [], error: "Agent 未配置比赛上下文", elapsedMs: 0 };
    }
    const started = Date.now();
    const actions = [];
    const captured = [];
    const targets = this._ctx.targets;
    for (const target of targets) {
      if (!target || !target.baseUrl) continue;
      const action = await this._runAgainstAsync(target.baseUrl, submit);
      actions.push(action);
      if (action.flag) captured.push(action.flag);
      if (!action.ok) break;
    }
    return {
      ok: actions.every((a) => a.ok),
      actions,
      flagsCaptured: captured,
      elapsedMs: Date.now() - started,
      error: null,
    };
  }

  stop() {
    if (this._process && !this._process.killed) {
      this._process.kill("SIGTERM");
      setTimeout(() => {
        if (this._process && !this._process.killed) {
          this._process.kill("SIGKILL");
        }
      }, 5000);
    }
  }
}

class AgentManager {
  /**
   * @param {CustomCommandAdapter} adapter
   */
  constructor(adapter) {
    this.adapter = adapter;
    /** @type {AgentResult[]} */
    this._results = [];
    this._running = false;
  }

  configure(matchConfig, roomStatus) {
    const ctx = makeContext(matchConfig, roomStatus);
    this.adapter.configure(ctx);
  }

  /**
   * @param {FlagSubmitter} [submit]
   * @returns {AgentResult}
   */
  runAttack(submit) {
    this._running = true;
    const result = this.adapter.run(submit);
    this._results.push(result);
    this._running = false;
    return result;
  }

  /**
   * Async version — does NOT block the event loop.
   * @param {FlagSubmitter} [submit]
   * @returns {Promise<AgentResult>}
   */
  async runAttackAsync(submit) {
    this._running = true;
    const result = await this.adapter.runAsync(submit);
    this._results.push(result);
    this._running = false;
    return result;
  }

  stop() {
    this.adapter.stop();
    this._running = false;
  }

  get running() {
    return this._running;
  }

  get lastResult() {
    return this._results.length > 0 ? this._results[this._results.length - 1] : null;
  }
}

/**
 * Split agent output into natural-language activity steps.
 * Used by the Electron main process to report per-step AGENT_ACTIVITY events.
 */
function parseActivitySteps(output, targetUrl, ok, flag) {
  if (!output) return [];
  const steps = [];

  const chunks = output
    .split(/\n\n+/)
    .flatMap(para => {
      const trimmed = para.trim();
      if (!trimmed) return [];
      if (trimmed.length > 200) {
        return trimmed.split(/(?<=[.!?。！？])\s+/).filter(Boolean);
      }
      return [trimmed];
    })
    .map(s => s.trim().replace(/[\x00-\x08\x0B\x0C\x0E-\x1F]/g, "").replace(/\s+/g, " "))
    .filter(s => s.length > 15)
    .slice(0, 25);

  for (const chunk of chunks) {
    const isError = /error|fail|refused|denied|timeout/i.test(chunk);
    steps.push({ desc: chunk.slice(0, 200), ok: !isError });
  }

  if (flag) {
    steps.push({ desc: `Flag captured: ${flag}`, ok: true });
  }

  return steps;
}

module.exports = {
  makeContext,
  contextEnv,
  extractFlags,
  expandTemplate,
  sanitizeCommand,
  commandNeedsShell,
  quoteShellArg,
  prepareSpawnArgv,
  parseActivitySteps,
  CustomCommandAdapter,
  AgentManager,
  FLAG_PATTERN,
};
